sample/weighted.go

```diff
@@ -0,0 +1,133 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import (
+	"math/rand"
+
+	"github.com/gonum/floats"
+)
+
+// Weighted provides sampling without replacement from a collection of items with
+// non-uniform probability.
+type Weighted struct {
+	weights []float64
+	// heap is a weight heap.
+	//
+	// It keeps a heap-organised sum of remaining
+	// index weights that are available to be taken
+	// from.
+	//
+	// Each elements holds the sum of weights for
+	// the corresponding index, plus the sum of
+	// of its children's weights; the children
+	// of an element i can be found at positions
+	// 2*(i+1)-1 and 2*(i+1). The root of the
+	// weight heap is at element 0.
+	//
+	// See comments in container/heap for an
+	// explanation of the layout of a heap.
+	heap []float64
+	src  *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. If src is nil, rand.Rand is
+// used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
+	s := Weighted{
+		weights: make([]float64, len(w)),
+		heap:    make([]float64, len(w)),
+	}
+	s.ReweightAll(w)
+	return s
+}
+
+// Len returns the number of items held by the Weighted, including items
+// already taken.
+func (s Weighted) Len() int { return len(s.weights) }
+
+// Take returns an index from the Weighted with probability proportional
+// to the weight of the item. The weight of the item is then set to zero.
+// Take returns false if there are no items remaining.
+func (s Weighted) Take() (idx int, ok bool) {
```

Comment:
On reflection, I think the first option is better. If we take the second, we really need to throw away the Reweight method or keep careful track of weights, because the user can arbitrarily re-weight items, potentially invalidating our assumptions that the maximum held is actually a maximum. Just saying "Don't be silly." seems reasonable.